### PR TITLE
improve: reduce unnecessary cast

### PR DESCRIPTION
**Pull request by ouyuanning**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:
for expr : uint32Col = 10
old rule： we will build plan like:  cast(uint32Col as int64) = 10
new rule:  if 10(const_expr)  between uint32， then plan like: uint32Col = cast(10 as uint32)

### CODE REVIEW

**Review by aunjgr — APPROVED** [review]

File: pkg/sql/colexec/expr_eval.go [hunk 1]
```diff
@@ -41,6 +47,66 @@ var (
 	constTimestampType  = types.Type{Oid: types.T_timestamp}
 )
 
+func getConstVec(proc *process.Process, expr *plan.Expr, length int) (*vector.Vector, error) {
+	var vec *vector.Vector
+	t := expr.Expr.(*plan.Expr_C)
+	if t.C.GetIsnull() {
+		if types.T(expr.Typ.GetId()) == types.T_any {
+			vec = vector.NewConstNull(types.Type{Oid: types.T(expr.Typ.GetId())}, length)
```
Comment (fengttt) [inline pkg/sql/colexec/expr_eval.go:55]: I am still not sure what is a T_any.   Is T_any only used to indicate a typeless NULL?

File: pkg/sql/plan/base_binder.go [hunk 2]
```diff
@@ -923,6 +927,40 @@ func bindFuncExprImplByPlanExpr(name string, args []*Expr) (*plan.Expr, error) {
 			args[0].Typ = makePlan2Type(&returnType)
 		}
 	}
+
+	// rewrite some cast rule:  expr:  int32Col > 10,
+	// old rule: cast(int32Col as int64) >10 ,   new rule: int32Col > (cast 10 as int32)
+	switch name {
+	case "=", "<", "<=", ">", ">=", "<>":
+		// if constant's type higher than column's type
+		// and constant's value in range of column's type, then no cast was needed
```
Comment (fengttt) [inline pkg/sql/plan/base_binder.go:936]: The right optimization rule:
if constant type's value is in range, then no cast was needed.
if constant type's value is not in range, then all these comparisons should be folded to a const true or false (or changed to is/isnot null).

File: pkg/sql/colexec/expr_eval.go (hunk 1 above)
Comment (ouyuanning) [reply pkg/sql/colexec/expr_eval.go:55]: select * from t where c1 > ? and c2 < @var 
the ? & var is T_any also.   

and some case I don't know.

File: pkg/sql/plan/base_binder.go (hunk 2 above)
Comment (ouyuanning) [reply pkg/sql/plan/base_binder.go:936]: @aunjgr give the same suggestion. 
but that will do the rewrite for the function one by one. 
 "=" 's rule is different from ">"....  Maybe we can optimize it next time.

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**